openedx/core/djangoapps/user_api/views.py

```diff
@@ -345,10 +350,63 @@ def post(self, request):
             }
             return JsonResponse(errors, status=400)
 
+        password = data.get("password")
+        password_copy = data.get("password_copy")
+        if password and password_copy and password != password_copy:
+            errors = {
+                'password_copy': [{"user_message": "Passwords don't match."}]
+            }
+            return JsonResponse(errors, status=400)
+
         response = JsonResponse({"success": True})
         set_logged_in_cookies(request, response, user)
+
+        if getattr(settings, 'REGISTRATION_EMAIL_FULL_VERIFICATION', None) is not None:
+            email = data.get("email")
+            error = self._verify_email_really_exists(email)
+            if error:
+                errors = {
+                    'email': [{"user_message": error}]
+                }
+                return JsonResponse(errors, status=400)
+
         return response
 
+    def _verify_email_really_exists(self, email):
+        """Check if a email really exists,
+        returns None if success, otherwise an error message
+        """
+        mx_record = None
+        try:
+            domain = email.rsplit('@', 1)[-1]
+            records = dns.resolver.query(domain, 'MX')
+            mx_record = records[0].exchange
+            mx_record = str(mx_record)
+        except (BaseException, StandardError):
+            return u"Email domain '{domain}' doesn't exist.".format(domain=domain)
+
+        try:
+            # Get local server hostname
+            host = socket.gethostname()
```

Comment:
For this block of logic to validate the status and existence of the email box on the domain, I want to reiterate my concerns expressed in the previous code block and more:
1. We are exposing the internal hostnames of the server to any email box host. This is not desired. We should not expose any internal hostnames.